File: packages/android/src/main/kotlin/com/rarimo/unforgettable/LocationHash.kt
package com.rarimo.unforgettable

import java.net.URLEncoder
import java.net.URLDecoder

/**
 * Parameters for Unforgettable location hash
 */
data class UnforgettablePathParams(
    val dataTransferId: String,
    val encryptionPublicKey: String,
    val factors: List<RecoveryFactor>,
    val walletAddress: String? = null,
    val group: String? = null,
    val customParams: Map<String, String>? = null
)

/**
 * Keys used in URL parameters
 */
private enum class ParamsKey(val value: String) {
    DATA_TRANSFER_ID("id"),
    ENCRYPTION_PUBLIC_KEY("epk"),
    FACTORS("f"),
    WALLET_ADDRESS("wa"),
    GROUP("g")
}

private val ALL_PARAMS_KEYS = ParamsKey.values().map { it.value }

/**
 * Errors that can occur when parsing location hash
 */
sealed class LocationHashError : Exception() {
    object InvalidParameters : LocationHashError()
    data class InvalidFactorId(val factorId: String) : LocationHashError()
    data class NonIntegerFactor(val factor: String) : LocationHashError()
}

/**
 * Composes a location hash from parameters
 * @param params The parameters to encode
 * @return A location hash string starting with #
 */
fun composeUnforgettableLocationHash(params: UnforgettablePathParams): String {
    val queryParams = mutableListOf<String>()
    
    queryParams.add("${ParamsKey.DATA_TRANSFER_ID.value}=${URLEncoder.encode(params.dataTransferId, "UTF-8")}")
    queryParams.add("${ParamsKey.ENCRYPTION_PUBLIC_KEY.value}=${URLEncoder.encode(params.encryptionPublicKey, "UTF-8")}")
    queryParams.add("${ParamsKey.FACTORS.value}=${params.factors.joinToString(",") { it.value.toString() }}")
    
    params.walletAddress?.let {
        queryParams.add("${ParamsKey.WALLET_ADDRESS.value}=${URLEncoder.encode(it, "UTF-8")}")
    }
    
    params.group?.let {
        queryParams.add("${ParamsKey.GROUP.value}=${URLEncoder.encode(it, "UTF-8")}")
    }
    
    params.customParams?.forEach { (key, value) ->
        if (!ALL_PARAMS_KEYS.contains(key)) {
            queryParams.add("${URLEncoder.encode(key, "UTF-8")}=${URLEncoder.encode(value, "UTF-8")}")
        }
    }
    
    return "#${queryParams.joinToString("&")}"
}

/**
 * Parses a location hash into parameters
 * @param hash The location hash string (with or without leading #)
 * @return The parsed parameters
 * @throws LocationHashError if parsing fails
 */
fun parseUnforgettableLocationHash(hash: String): UnforgettablePathParams {
    val rawParams = if (hash.startsWith("#")) hash.substring(1) else hash
    
    val params = rawParams.split("&")
        .mapNotNull { param ->
            val parts = param.split("=", limit = 2)
            if (parts.size == 2) parts[0] to URLDecoder.decode(parts[1], "UTF-8")
            else null
        }
        .toMap()
    
    val customParams = params.filterKeys { !ALL_PARAMS_KEYS.contains(it) }
    
    val dataTransferId = params[ParamsKey.DATA_TRANSFER_ID.value]
    val encryptionPublicKey = params[ParamsKey.ENCRYPTION_PUBLIC_KEY.value]
    
    if (dataTransferId.isNullOrEmpty() || encryptionPublicKey.isNullOrEmpty()) {
        throw LocationHashError.InvalidParameters
    }
    
    val factors = parseFactors(params[ParamsKey.FACTORS.value] ?: "")
    val walletAddress = params[ParamsKey.WALLET_ADDRESS.value]
    val group = params[ParamsKey.GROUP.value]
    
    return UnforgettablePathParams(
        dataTransferId = dataTransferId,
        encryptionPublicKey = encryptionPublicKey,
        factors = factors,
        walletAddress = walletAddress,
        group = group,
        customParams = if (customParams.isEmpty()) null else customParams
    )
}

// Private helper functions

private fun parseFactors(rawFactors: String): List<RecoveryFactor> {
    if (rawFactors.isEmpty()) {
        return emptyList()
    }
    
    val factorStrings = rawFactors.split(",").filter { it.isNotEmpty() }
    val factors = mutableListOf<RecoveryFactor>()
    
    for (rawFactor in factorStrings) {
        val num = rawFactor.toIntOrNull()
            ?: throw LocationHashError.NonIntegerFactor(rawFactor)
        
        val factor = RecoveryFactor.fromValue(num)
            ?: throw LocationHashError.InvalidFactorId(rawFactor)
        
        factors.add(factor)
    }
    
    return factors
}

File: packages/ios/Sources/UnforgettableSDK/LocationHash.swift
import Foundation

/// Parameters for Unforgettable location hash
public struct UnforgettablePathParams {
    public let dataTransferId: String
    public let encryptionPublicKey: String
    public let factors: [RecoveryFactor]
    public let walletAddress: String?
    public let group: String?
    public let customParams: [String: String]?
    
    public init(
        dataTransferId: String,
        encryptionPublicKey: String,
        factors: [RecoveryFactor],
        walletAddress: String? = nil,
        group: String? = nil,
        customParams: [String: String]? = nil
    ) {
        self.dataTransferId = dataTransferId
        self.encryptionPublicKey = encryptionPublicKey
        self.factors = factors
        self.walletAddress = walletAddress
        self.group = group
        self.customParams = customParams
    }
}

/// Keys used in URL parameters
private enum ParamsKey: String {
    case dataTransferId = "id"
    case encryptionPublicKey = "epk"
    case factors = "f"
    case walletAddress = "wa"
    case group = "g"
}

private let ALL_PARAMS_KEYS: [String] = [
    ParamsKey.dataTransferId.rawValue,
    ParamsKey.encryptionPublicKey.rawValue,
    ParamsKey.factors.rawValue,
    ParamsKey.walletAddress.rawValue,
    ParamsKey.group.rawValue
]

/// Errors that can occur when parsing location hash
public enum LocationHashError: Error {
    case invalidParameters
    case invalidFactorId(String)
    case nonIntegerFactor(String)
}

/// Composes a location hash from parameters
/// - Parameter params: The parameters to encode
/// - Returns: A location hash string starting with #
public func composeUnforgettableLocationHash(params: UnforgettablePathParams) -> String {
    var components = URLComponents()
    components.queryItems = []
    
    components.queryItems?.append(URLQueryItem(
        name: ParamsKey.dataTransferId.rawValue,
        value: params.dataTransferId
    ))
    
    components.queryItems?.append(URLQueryItem(
        name: ParamsKey.encryptionPublicKey.rawValue,
        value: params.encryptionPublicKey
    ))
    
    let factorsString = params.factors.map { String($0.rawValue) }.joined(separator: ",")
    components.queryItems?.append(URLQueryItem(
        name: ParamsKey.factors.rawValue,
        value: factorsString
    ))
    
    if let walletAddress = params.walletAddress {
        components.queryItems?.append(URLQueryItem(
            name: ParamsKey.walletAddress.rawValue,
            value: walletAddress
        ))
    }
    
    if let group = params.group {
        components.queryItems?.append(URLQueryItem(
            name: ParamsKey.group.rawValue,
            value: group
        ))
    }
    
    if let customParams = params.customParams {
        for (key, value) in customParams {
            if !ALL_PARAMS_KEYS.contains(key) {
                components.queryItems?.append(URLQueryItem(
                    name: key,
                    value: value
                ))
            }
        }
    }
    
    return "#" + (components.query ?? "")
}

/// Parses a location hash into parameters
/// - Parameter hash: The location hash string (with or without leading #)
/// - Returns: The parsed parameters
/// - Throws: LocationHashError if parsing fails
public func parseUnforgettableLocationHash(_ hash: String) throws -> UnforgettablePathParams {
    let rawParams = hash.hasPrefix("#") ? String(hash.dropFirst()) : hash
    
    guard let components = URLComponents(string: "?" + rawParams),
          let queryItems = components.queryItems else {
        throw LocationHashError.invalidParameters
    }
    
    var params: [String: String] = [:]
    var customParams: [String: String] = [:]
    
    for item in queryItems {
        if let value = item.value {
            params[item.name] = value
            if !ALL_PARAMS_KEYS.contains(item.name) {
                customParams[item.name] = value
            }
        }
    }
    
    guard let dataTransferId = params[ParamsKey.dataTransferId.rawValue],
          let encryptionPublicKey = params[ParamsKey.encryptionPublicKey.rawValue],
          !dataTransferId.isEmpty,
          !encryptionPublicKey.isEmpty else {
        throw LocationHashError.invalidParameters
    }
    
    let factors = try parseFactors(params[ParamsKey.factors.rawValue] ?? "")
    let walletAddress = params[ParamsKey.walletAddress.rawValue]
    let group = params[ParamsKey.group.rawValue]
    
    return UnforgettablePathParams(
        dataTransferId: dataTransferId,
        encryptionPublicKey: encryptionPublicKey,
        factors: factors,
        walletAddress: walletAddress,
        group: group,
        customParams: customParams.isEmpty ? nil : customParams
    )
}

// MARK: - Private Helpers

private func parseFactors(_ rawFactors: String) throws -> [RecoveryFactor] {
    if rawFactors.isEmpty {
        return []
    }
    
    let factorStrings = rawFactors.split(separator: ",").map(String.init).filter { !$0.isEmpty }
    var factors: [RecoveryFactor] = []
    
    for rawFactor in factorStrings {
        guard let num = Int(rawFactor) else {
            throw LocationHashError.nonIntegerFactor(rawFactor)
        }
        
        guard let factor = RecoveryFactor(rawValue: num) else {
            throw LocationHashError.invalidFactorId(rawFactor)
        }
        
        factors.append(factor)
    }
    
    return factors
}
